### PR TITLE
[CUDA][cuBLAS] Optionally autotune `should_fold` in matmuls

### PR DESCRIPTION
**Pull request by eqy**
Adds `torch.set_matmul_autotune(True)` to support matmul autotuning as requested in #118774

Ostensibly this also supports the CPU use case but it might be unnecessary...

benchmark script and results:
```
import torch
import time
import pandas as pd
torch.manual_seed(0)
dim_max = 2048
warmup = 3
iters = 10
print(torch.get_matmul_autotune())
torch.set_matmul_autotune(True)
print(torch.get_matmul_autotune())

data = {'b': [], 'k': [], 'm': [], 'autotune': [], 'default': [], 'speedup': []}


for _ in range(50):
    b = int(torch.randint(dim_max, (1,)))
    k = int(torch.randint(dim_max, (1,)))
    m = int(torch.randint(dim_max, (1,)))
    tensor1 = torch.randn(k, k, device='cuda', dtype=torch.half, requires_grad=True)
    tensor2 = torch.randn(b, m, k, device='cuda', dtype=torch.half).transpose(1, 2).contiguous()

    data['b'].append(b)
    data['k'].append(k)
    data['m'].append(m)

    for autotune in False, True:
        torch.set_matmul_autotune(autotune)
        for _ in range(warmup):
            torch.matmul(tensor1, tensor2)
        torch.cuda.synchronize()
        t1 = time.perf_counter()
        for _ in range(iters):
            torch.matmul(tensor1, tensor2)
        torch.cuda.synchronize()
        t2 = time.perf_counter()
        lat = (t2 - t1)/iters
        if autotune:
            data['autotune'].append(lat)
        else:
            data['default'].append(lat)
    data['speedup'].append(data['default'][-1]/data['autotune'][-1])
df = pd.DataFrame.from_dict(data)
print(df.sort_values(by=['speedup'], ascending=False))
```
```
       b     k     m  autotune   default   speedup
31  1106    91  1920  0.000828  0.003731  4.504789
49   917   201   383  0.000538  0.001790  3.328740
3   1747   277  1778  0.007975  0.023111  2.897854
29    57   291  1894  0.000305  0.000873  2.861024
7    551    87   174  0.000057  0.000142  2.506835
33   908   633  1962  0.016720  0.039514  2.363268
16   910   423  1312  0.006649  0.015457  2.324661
0    684   559  1653  0.009180  0.021103  2.298807
4   1828   599  1094  0.018014  0.041221  2.288197
25   273   335   388  0.000452  0.001024  2.264646
32   398   611  1589  0.006029  0.013485  2.236690
11   976   755  1733  0.020939  0.046366  2.214362
6    314   705  1510  0.005496  0.012041  2.190730
34    84   203   324  0.000063  0.000139  2.185001
37   972   868  1204  0.014627  0.031652  2.163959
5   1496   600  1420  0.013428  0.028574  2.127922
8   1624   849  1701  0.044462  0.093258  2.097490
1   1216   835   763  0.014208  0.029344  2.065296
26  1641  1066  1466  0.043851  0.090297  2.059178
15   797   659  1171  0.010925  0.022443  2.054242
40   791   719   909  0.009571  0.019172  2.003002
20   675   510  1483  0.006201  0.012312  1.985313
48   973   368  2023  0.007667  0.015195  1.981900
44  1221  1630   256  0.010805  0.020882  1.932552
23  1956  1077  1925  0.095383  0.183523  1.924075
14  2008  1201   755  0.047141  0.090151  1.912375
30  1143   779   430  0.008053  0.014681  1.823037
39  1684  1251   954  0.054064  0.097876  1.810381
19   714   244   151  0.000316  0.000563  1.779779
17  1985  1289   697  0.053919  0.092060  1.707401
45  1905  1714  1316  0.113937  0.185914  1.631719
47   387  1634  1578  0.025715  0.041956  1.631606
12  2046  1871  1455  0.216917  0.348048  1.604519
28  1345  2023  1961  0.224436  0.355593  1.584383
2   1731  1383  1033  0.079691  0.125319  1.572559
46  1954   816  1885  0.059301  0.079783  1.345406
41   373  1877   560  0.018621  0.024595  1.320788
24  1574  1512   756  0.038369  0.050582  1.318313
35   774   964  1071  0.019876  0.026161  1.316207
18   639   544   543  0.003104  0.003927  1.265235
42  1329  1605   169  0.016360  0.020254  1.237986
38  2024  1870  1167  0.178119  0.205499  1.153719
13   448   850    99  0.000878  0.000984  1.121591
9    537   845    72  0.001097  0.001174  1.070789
36   639  2036  1155  0.063162  0.067374  1.066683
21   882  1207    28  0.001258  0.001281  1.018017
22   802   128   128  0.000216  0.000216  1.003902
43  1699  1472  1119  0.079897  0.079596  0.996241
10   777  1940   115  0.007235  0.007140  0.986898
27   543   888   257  0.002863  0.002757  0.963044
```

Note that these numbers were collected on a "power-limited" H100, so YMMV

CC @IvanYashchuk @lezcano @xuzhao9 @soulitzer @ptrblck 

cc @ptrblck @csarofeen @xwang233

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
...
+};
+
+struct MatmulAutotuneCacheKeyWrapper : ParamsWrapper<MatmulAutotuneCacheKey> {
+  MatmulAutotuneCacheKeyWrapper() = default;
+
+  MatmulAutotuneCacheKeyWrapper(
+    const Tensor& tensor1,
+    const Tensor& tensor2,
+    bool should_fold) {
+    TORCH_INTERNAL_ASSERT(tensor1.sizes().size() <= AUTOTUNE_MAX_DIM);
```
Comment (eqy) [inline aten/src/ATen/native/LinearAlgebra.cpp:1934]: TODO: fail gracefully here (bail out of autotuning rather than crashing)

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -11,17 +11,20 @@
...
 #include <ATen/native/ReduceOps.h>
 #include <ATen/native/ReduceOpsUtils.h>
 #include <ATen/native/Resize.h>
 #include <ATen/native/mkldnn/Matmul.h>
+#include <ATen/native/utils/ParamsHash.h>
 #include <c10/core/GradMode.h>
 #include <c10/util/accumulate.h>
 #include <c10/util/irange.h>
 #include <variant>
+#include <iostream>
```
Comment (eqy) [inline aten/src/ATen/native/LinearAlgebra.cpp:27]: cleanup

**Comment by xuzhao9** [comment]
One question: does this option apply to both eager and torch.compile()?

**Comment by eqy** [comment]
> One question: does this option apply to both eager and torch.compile()?

I'm not familiar with the details here but `torch.compile` goes through the same `should_fold` path when dispatching to ATen, then yes it would also apply.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
   return result.fill_(self.vdot(other));
 }
 
-static bool should_fold(const Tensor& tensor1, const Tensor& tensor2) {
+#define AUTOTUNE_MAX_DIM 4
+
+struct MatmulAutotuneCacheKey {
+  int tensor1_dim[AUTOTUNE_MAX_DIM];
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebra.cpp:1917]: The type of elements of Tensor.sizes() is `int64_t`. You could also use a `DimVector` from `c10/util/DimVector.h`.

File: torch/__init__.py [hunk 4]
```diff
@@ -1065,6 +1066,22 @@ def set_float32_matmul_precision(precision: str) -> None:
...
+def get_matmul_autotune() -> builtins.bool:
+    r"""Returns the current value of the matrix multiplication autotuning flag. Refer to
+    :func:`torch.set_matmul_autotune` documentation for more details.
+    """
+    return _C._get_matmul_autotune()
+
+def set_matmul_autotune(enabled: bool) -> None:
+    r"""Sets the autotuning flag for matrix multiplications.
+
+    Certain matmuls can be folded into bmm (batched matrix multiply), and by default the decision
```
Comment (IvanYashchuk) [inline torch/__init__.py:1078]: If one of the arguments of batched matrix multiplication is N-dimensions with N >= 3 and the other argument is 2-dimensional then "should_fold" decides whether to fold the N-dimensional argument into 2-dimensional and call a single matrix multiplication (`torch.mm`).

**Comment by IvanYashchuk** [comment]
@lezcano, since you're best familiar with all the choices in the `LinearAlgebra.cpp:should_fold` function could please review this PR and make sure we're not breaking anything here?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 5]
```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
...
+  int tensor1_stride[AUTOTUNE_MAX_DIM];
+  int tensor2_dim[AUTOTUNE_MAX_DIM];
+  int tensor2_stride[AUTOTUNE_MAX_DIM];
+  bool should_fold;
+  ScalarType dtype;
+  bool is_cuda;
+  int device;
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:1924]: Push small types to the end so that they can be packed better
```suggestion
  ScalarType dtype;
  int device;
  bool is_cuda;
  bool should_fold;
```

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp:1934]: Using a Dimvector should do here

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 6]
```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
...
+  engine_cache.erase(key);
+  engine_cache.emplace(key, std::move(results));
+}
+
+};
+
+MatmulAutotuneCache<float, MatmulAutotuneCacheKeyWrapper> matmul_autotune_cache;
+
+static bool should_fold(const Tensor& tensor1, const Tensor& tensor2, MatmulAutotuneCacheKeyWrapper & key, bool& need_profiling, const bool autotune) {
+  if (autotune) {
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:1969]: This should be moved just before the line
```
  // t1->view(-1, t1->size(-1)) does not copy only when the first n-1 dimensions are contiguous
```
In all the other cases we are quite sure we want to fold.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 7]
```diff
@@ -2014,7 +2097,15 @@ static Tensor _matmul_impl(
                    : tensor1.unsqueeze(0).mm(tensor2).squeeze_(0);
   } else if (dim_tensor1 == 2 && dim_tensor2 == 2) {
     return has_out ? at::mm_out(out, tensor1, tensor2) : tensor1.mm(tensor2);
-  } else if (should_fold(tensor1, tensor2)) {
+  } else if (should_fold(tensor1, tensor2, key, need_profiling, autotune)) {
+    std::chrono::time_point<std::chrono::high_resolution_clock> start;
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2101]: A much cleaner way to implement this would be via a RAII object. This way youñ won't need to repeat the code in both branches, and you won't need to refactor the else branch.

**Review by lezcano — COMMENTED** [review]
As it stands, the PR is buggy as we may fold in cases when we can't fold (pushing down the profiling in `should_fold`  should fix this, but yeah). To give it a proper test to these changes, could you enable this feature by default, see what CI has to say?

Other than that, I really like the PR and this approach! Just one more comments:

- It'd be cool to make this persistent. Do we have precedent to of loading/storing this sort of info? cc @ezyang @albanD 

Regarding `torch.compile`, I think the better approach there would be to add this pattern to the autotuner we already have in inductor. cc @isuruf in case you want to take on this one.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 8]
```diff
@@ -1908,7 +1911,84 @@ Tensor& vdot_out(const Tensor& self, const Tensor& other, Tensor& result) {
...
+
+T* find(const KeyType& key) {
+  auto it = engine_cache.find(key);
+  if (it == engine_cache.end()) {
+    return nullptr;
+  }
+  return &(it->second);
+}
+
+void update(const KeyType& key, T& results) {
```
Comment (Skylion007) [inline aten/src/ATen/native/LinearAlgebra.cpp:1959]:
```suggestion
void update(const KeyType& key, T&& results) {
```
or
```suggestion
void update(const KeyType& key, T results) {
```
Unless I am missing something? I'm confused why you want a  to reference T here.

**Review by albanD — COMMENTED** [review]
I don't know that we do anything like this no. There are similar discussions for cudnn here: https://github.com/pytorch/pytorch/issues/88410
As discussed there, having a cache for such thing is non-trivial but maybe we can use similar logic to do both?

Also given how cuda-specific the code is, I would move it to `torch.backends.cuda.` for now. It is always easier to add a new global setting afterwards if people want to implement it.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (eqy) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Will go with `int64_t` as the params hashing complains that `DimVector` isn't POD

**Comment by eqy** [comment]
> I don't know that we do anything like this no. There are similar discussions for cudnn here: #88410 As discussed there, having a cache for such thing is non-trivial but maybe we can use similar logic to do both?
> 
> Also given how cuda-specific the code is, I would move it to `torch.backends.cuda.` for now. It is always easier to add a new global setting afterwards if people want to implement it.

As it stands there should be nothing CUDA-specific in the code now, all CUDA specific stuff has been moved into the timing function. If anything it would be more work to specialize on never turning autotuning on for the CPU case.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (eqy) [reply aten/src/ATen/native/LinearAlgebra.cpp]: (DimVector unfortunately doesn't seem to work at it isn't POD, a requirement of the hashing util)

**Comment by ezyang** [comment]
> It'd be cool to make this persistent. Do we have precedent to of loading/storing this sort of info?

I think the flash attention folks (cc @drisspg @cpuhrsch) autotuned a bunch of configs and then just checked them into the source code as a big table. I think that's a very practical way to do it if there are not too many.

**Comment by albanD** [comment]
> As it stands there should be nothing CUDA-specific in the code now, all CUDA specific stuff has been moved into the timing function. If anything it would be more work to specialize on never turning autotuning on for the CPU case.

Sure but do we have that running today? If there is no concrete plan to generalize it, I think my point still stands. We could do it is very different from we have a plan to do it.
If we do have a plan to do it, I would still move it out of the top level API to either `torch.backends.generic` or `torch.utils.perf_config` where we can put shared flags.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 6 above)
Comment (eqy) [reply aten/src/ATen/native/LinearAlgebra.cpp]: I did a bit of twiddling and found out what makes the biggest difference is moving the autotuning before 
```
2022   bool t2_requires_grad = tensor1_larger ? tensor2.requires_grad() : tensor1.requires_grad();
2023   if (t2_requires_grad && !has_out) {
```

Comment (lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp]: That doesn't make much sense as, in your tests, the tensors should not have requires_grad or out, no?

Comment (eqy) [reply aten/src/ATen/native/LinearAlgebra.cpp]: `tensor1` does have `requires_grad=True` in the test but has smaller `dim`, so I'm assuming some speedup comes from autotuning not folding when the heuristic would fold?

Comment (lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp]: If it has `requires_grad=True` you should benchmark forward + backward to make it realistic. Not folding in those cases would incur in a large tensor being materialised in the backward.

**Comment by eqy** [comment]
CPU failures are interesting given that autotuning is disabled there, also need to find a machine that reproduces the issue locally, as it doesn't seem to on a few AMD-based nodes that I've tried...

File: test/functorch/test_ops.py [hunk 9]
```diff
@@ -1960,6 +1978,10 @@ def f(x):
     @skipOps('TestOperators', 'test_vmapvjpvmap', {
         xfail('NumpyCubeNotComposableAutogradFunction'),  # Not composable
     })
+    @opsToleranceOverride('TestOperators', 'test_vmapjvpvjp', (
+        tol1('linalg.multi_dot',
+             {torch.float32: tol(atol=1e-04, rtol=1e-05)}),
```
Comment (eqy) [inline test/functorch/test_ops.py:1983]: Ah, missing condition for `cuda` only ?

**Comment by lezcano** [comment]
Being quite honest, the huge increase in the tolerances needed for this PR to pass seems quite concerning...

On a similar note, can you post the speed-ups given by this method taking https://github.com/pytorch/pytorch/pull/119117#discussion_r1483581778 into account?

**Comment by eqy** [comment]
> Being quite honest, the huge increase in the tolerances needed for this PR to pass seems quite concerning...
> 
> On a similar note, can you post the speed-ups given by this method taking [#119117 (comment)](https://github.com/pytorch/pytorch/pull/119117#discussion_r1483581778) into account?

One interesting thing is that virtually all of the test failures go way in my setup (A100/H100) when I disable TF32, so I wonder if it's a case of the folding picking kernels that exacerbate TF32 precision issues.

I'll take a look at the backward perf as well.

**Comment by eqy** [comment]
> The issue with the backward op is that you cannot benchmark it at that point. For that part, you need to have a static heuristic. If the current heuristic can be improved, then that'd probably belong in a different PR.

Can we infer the forward decision during the backward pass and naively update timing data during the backward as the total forward + backward time?

**Comment by lezcano** [comment]
You could benchmark both forward and backward in the forward, but that looks like as bit of a push. 

I think it'd be best to just put together a static table with some cutoff shapes deciding between one path or the other. We do this all throughout the linalg module really
https://github.com/pytorch/pytorch/blob/3629287151245c98f4d04b86729ca991cfdbbe3e/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp#L2512-L2535
